### ISSUE TITLE
Handle circular relation tuples

### ISSUE THREAD

**Issue by zepatrik** [ory/keto#411]
**Describe the bug**

Currently, both the check and expand engines don't handle circular relation tuples but resolve references indefinitely.

**Reproducing the bug**

```go
t.Run("case=circular tuples", func(t *testing.T) {
		sendlingerTor, odeonsplatz, centralStation, connected, namesp := "Sendlinger Tor", "Odeonsplatz", "Central Station", "connected", "munich transport"

		reg := newDepsProvider(t, []*namespace.Namespace{{Name: namesp}})

		require.NoError(t, reg.RelationTupleManager().WriteRelationTuples(context.Background(), []*relationtuple.InternalRelationTuple{
			{
				Namespace: namesp,
				Object:    sendlingerTor,
				Relation:  connected,
				Subject: &relationtuple.SubjectSet{
					Namespace: namesp,
					Object:    odeonsplatz,
					Relation:  connected,
				},
			},
			{
				Namespace: namesp,
				Object:    odeonsplatz,
				Relation:  connected,
				Subject: &relationtuple.SubjectSet{
					Namespace: namesp,
					Object:    centralStation,
					Relation:  connected,
				},
			},
			{
				Namespace: namesp,
				Object:    centralStation,
				Relation:  connected,
				Subject: &relationtuple.SubjectSet{
					Namespace: namesp,
					Object:    sendlingerTor,
					Relation:  connected,
				},
			},
		}...))

		e := check.NewEngine(reg)

		stations := []string{sendlingerTor, odeonsplatz, centralStation}
		for _, start := range stations {
			for _, end := range stations {
				res, err := e.SubjectIsAllowed(context.Background(), &relationtuple.InternalRelationTuple{
					Namespace: namesp,
					Object:    start,
					Relation:  connected,
					Subject: &relationtuple.SubjectID{
						ID: end,
					},
				})
				require.NoError(t, err)
				assert.True(t, res)
			}
		}
	})
```
never terminates.

**Expected behavior**

We should either:
- disallow circular relation tuples (will have to detect them then)
- handle circular tuples

I think we will have to add some kind of cache anyway (#312). It could be used as well to keep track of already resolved references.

**Comment by aravindarc** [comment]
I have written a small utility function that can find out cycles when resolving a check, it looks something like this

```golang
func checkVisited(ctx context.Context, current *relationtuple.InternalRelationTuple) (ctxt context.Context, wasAlreadyVisited bool) {
	visitedMap, ok := ctx.Value("visitedMap").(map[string]bool)
	if !ok {
		// for the first time initialize the map
		visitedMap = make(map[string]bool)
		visitedMap[current.String()] = true
		return context.WithValue(ctx, "visitedMap", visitedMap), false
	}

	// check if current node was already visited
	if visitedMap[current.String()] {
		return ctx, true
	}

	// set current entry to visited
	visitedMap[current.String()] = true
	
	return context.WithValue(
		ctx,
		"visitedMap",
		visitedMap,
	), false
}
```
And I called this function below this line https://github.com/ory/keto/blob/e5b76b35c495add911b6c2c61c4f51716ac64a3f/internal/check/engine.go#L37

like 

```go
for _, sr := range rels {
	ctx, wasAlreadyVisited := checkVisited(ctx, sr)
	if wasAlreadyVisited {
		return false, errors.New("Cycle found at " + sr.String())
	}
        ...
}
```

I checked this against your test case, it seems to be finding out the cycle, what do you think

**Comment by aravindarc** [comment]
Yeah sure, I just wanted to get your opinion before I raise a PR. Will handle both check and expand in that

**Comment by zepatrik** [comment]
Nice, looking forward for that PR :eyes: :wink:

**Comment by zepatrik** [comment]
closed in #623
